### PR TITLE
Add support for FOTA combined image

### PR DESCRIPTION
**Pull request by davidsaada**
This PR starts the support for FOTA combined image for edge. This now includes the rootfs installation and verification callbacks, where the rest of rootfs callbacks (rollback, finalize) and all bootloader callbacks still need implementation.

This includes the following:
- Replace `fota_install_callbacks` module with `fota_app_callbacks`, which includes all FOTA related callbacks (including the ones previously used for FOTA).
- Use external defines for install script directory and file.
- Add relevant defines in makefile if `FOTA_COMBINED_IMAGE_SUPPORT` is set.

### CODE REVIEW

File: cmake/edge_configure.cmake [hunk 1]
```diff
@@ -109,9 +109,23 @@ if (${FIRMWARE_UPDATE})
   MESSAGE ("Enabling firmware update for Edge")
 
   if (FOTA_ENABLE)
-    execute_process ( COMMAND cp ${CMAKE_CURRENT_SOURCE_DIR}/fota/fota_install_callback.c ${CMAKE_CURRENT_SOURCE_DIR}/lib/mbed-cloud-client/fota/ )
+    execute_process ( COMMAND cp ${CMAKE_CURRENT_SOURCE_DIR}/fota/fota_app_callbacks.c ${CMAKE_CURRENT_SOURCE_DIR}/lib/mbed-cloud-client/fota/ )
+    if (NOT DEFINED FOTA_SCRIPT_DIR)
+      SET (FOTA_SCRIPT_DIR \"/opt/pelion\")
+    endif()
+    if (NOT DEFINED FOTA_INSTALL_MAIN_SCRIPT)
+      SET (FOTA_SCRIPT_DIR \"fota_update_activate.sh\")
```
Comment (petedyerarm) [inline cmake/edge_configure.cmake:117]:
```suggestion
      SET (FOTA_INSTALL_MAIN_SCRIPT \"fota_update_activate.sh\")
```

File: fota/fota_app_callbacks.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            FOTA_TRACE_ERROR("shell could not be run");
+        } else {
+            FOTA_TRACE_ERROR("Installation failed. Result of running command is %d", WEXITSTATUS(rc));
+        }
+    }
+    return ret;
+}
+
+
+int verify_ostree_delta_update(const char *commit)
```
Comment (petedyerarm) [inline fota/fota_app_callbacks.c:59]: If this is this only called when MBED_CLOUD_CLIENT_FOTA_SUB_COMPONENT_SUPPORT == 1 then it should only be included in that case.

**Review by petedyerarm — CHANGES_REQUESTED** [review]

File: fota/fota_app_callbacks.c (hunk 2 above)
Comment (davidsaada) [reply fota/fota_app_callbacks.c:59]: Good point for discussion. I actually want to use it also for the regular case. This is something we wanted to do back then, but we didn't.

File: cmake/edge_configure.cmake (hunk 1 above)
Comment (davidsaada) [reply cmake/edge_configure.cmake:117]: Will fix, but why not `FOTA_INSTALL_MAIN_SCRIPT_DIR`?

File: cmake/edge_configure.cmake [hunk 3]
```diff
@@ -109,9 +109,23 @@ if (${FIRMWARE_UPDATE})
...
+    endif()
+
     add_definitions(-DFOTA_DEFAULT_APP_IFS=1)
     add_definitions(-DTARGET_LIKE_LINUX=1)
+    add_definitions(-DFOTA_SCRIPT_DIR=${FOTA_SCRIPT_DIR})
+    add_definitions(-DFOTA_INSTALL_MAIN_SCRIPT=${FOTA_INSTALL_MAIN_SCRIPT})
+    if (FOTA_COMBINED_IMAGE_SUPPORT)
```
Comment (ygoyal18) [inline cmake/edge_configure.cmake:124]: @davidsaada if the user wants this feature then they have to build with `-DFIRMWARE_UPDATE=ON` `-DFOTA_ENABLE=ON` and `-DFOTA_COMBINED_IMAGE_SUPPORT=ON`, correct?

File: fota/fota_app_callbacks.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
+// ----------------------------------------------------------------------------
+// Copyright 2020 ARM Ltd.
```
Comment (ygoyal18) [inline fota/fota_app_callbacks.c:2]: I think it should be 2021 Pelion IoT Ltd.

File: cmake/edge_configure.cmake [hunk 5]
```diff
@@ -109,9 +109,23 @@ if (${FIRMWARE_UPDATE})
...
+      SET (FOTA_SCRIPT_DIR \"fota_update_activate.sh\")
+    endif()
+
     add_definitions(-DFOTA_DEFAULT_APP_IFS=1)
     add_definitions(-DTARGET_LIKE_LINUX=1)
+    add_definitions(-DFOTA_SCRIPT_DIR=${FOTA_SCRIPT_DIR})
+    add_definitions(-DFOTA_INSTALL_MAIN_SCRIPT=${FOTA_INSTALL_MAIN_SCRIPT})
+    if (FOTA_COMBINED_IMAGE_SUPPORT) 
+      add_definitions(-DMBED_CLOUD_CLIENT_FOTA_SUB_COMPONENT_SUPPORT=1)
+      add_definitions(-DFOTA_CUSTOM_PLATFORM=1)
```
Comment (ygoyal18) [inline cmake/edge_configure.cmake:126]: @davidsaada what is the purpose of this flag?

File: cmake/edge_configure.cmake (hunk 3 above)
Comment (davidsaada) [reply cmake/edge_configure.cmake:124]: Yes. I will create a separate PR for meta-mbed-edge that will include these (currently only for uz boards).

File: cmake/edge_configure.cmake (hunk 5 above)
Comment (davidsaada) [reply cmake/edge_configure.cmake:126]: The way FOTA works, any platform specific hooks required in cases of init or update start/finish/abort require this flag to be set. In our case, we need to have `fota_platform_init_hook` to create the combined update handlers.

**Comment by davidsaada** [comment]
@ygoyal18 @petedyerarm Comments addressed, please re-review.

**Comment by ygoyal18** [comment]
I am able to recreate the PR check error on my local machine. For some reason, when you compile parsec_se_driver with new version of cargo 1.55.0, its failing to link the library. Its working with older version of cargo like 1.52.0.

**Comment by petedyerarm** [comment]
> I am able to recreate the PR check error on my local machine. For some reason, when you compile parsec_se_driver with new version of cargo 1.55.0, its failing to link the library. Its working with older version of cargo like 1.52.0.

What determines the version of cargo to be used?

**Review by petedyerarm — APPROVED** [review]
LGTM

**Comment by davidsaada** [comment]
@ygoyal18 Is it OK to merge this PR, as CI errors are unrelated?

**Comment by ygoyal18** [comment]
Yes, I agree that these changes are not responsible for PR checks failure. I am trying to find a quick solution to the problem, looks like some environment configuration issue. If not resolved in 24 hours, I will merge this PR. 

@petedyerarm This is where its downloading the latest available cargo and rustc - https://github.com/PelionIoT/edge-ci/blob/98cd81fc016f0ea3ca283a675805c919218a2409/vars/repoMbedEdge.groovy#L82 . Quick solution would be to lock down the version to 1.52.0.

**Review by prakharbhatt007 — APPROVED** [review]
Looks good.

**Pull request merged**